packages/rehype-expressive-code/src/index.ts

```diff
@@ -114,39 +165,72 @@ export type RehypeExpressiveCodeDocument = {
 	data: Record<string, unknown> | undefined
 }
 
-export type RehypeExpressiveCodeRenderer = {
-	ec: ExpressiveCode
+type RehypeExpressiveCodeEngineRenderer<T extends ExpressiveCodeEngine> = {
+	ec: T
 	baseStyles: string
 	themeStyles: string
 	jsModules: string[]
 }
 
+export type RehypeExpressiveCodeCoreRenderer = RehypeExpressiveCodeEngineRenderer<ExpressiveCodeCore>
+export type RehypeExpressiveCodeRenderer = RehypeExpressiveCodeEngineRenderer<ExpressiveCode>
+
 /**
  * Creates an `ExpressiveCode` instance using the given `options`,
  * including support to load themes bundled with Shiki by name.
  *
  * Returns the created `ExpressiveCode` instance together with the base styles and JS modules
  * that should be added to every page.
  */
-export async function createRenderer(options: RehypeExpressiveCodeOptions = {}): Promise<RehypeExpressiveCodeRenderer> {
+export async function createRenderer(options: RehypeExpressiveCodeOptions): Promise<RehypeExpressiveCodeRenderer> {
+	return createRendererCommon(options, { ctor: ExpressiveCode, loadTheme: loadShikiTheme })
+}
+
+/**
+ * Creates an `ExpressiveCodeCore` instance using the given `options`.
+ *
+ * Returns the created `ExpressiveCodeCore` instance together with the base styles and JS modules
+ * that should be added to every page.
+ */
+export async function createRendererCore<T extends string = never>(options: RehypeExpressiveCodeCoreOptions<T>): Promise<RehypeExpressiveCodeCoreRenderer> {
+	// typescript can't infer T here so it requires loadTheme but we do not provide a default loadTheme for Core,
+	// customLoadTheme must be used if T extends string so we must cast
+	return createRendererCommon(options, { ctor: ExpressiveCodeCore } as CreateRendererDefaultOptions<ExpressiveCodeCore, T>)
+}
+
+type CreateRendererDefaultOptions<E extends ExpressiveCodeCore, T extends string = never> = {
+	ctor: new (...args: ConstructorParameters<typeof ExpressiveCodeEngine>) => E
+} & ([T] extends [never] ? object : { loadTheme: LoadTheme<T> }) // loadTheme must be provided when T extends string
+
+async function createRendererCommon<E extends ExpressiveCodeEngine, T extends string = never>(
+	options: RehypeExpressiveCodeCommonOptions<ExpressiveCodeCoreConfig, E, T>,
+	defaultOptions: CreateRendererDefaultOptions<E, T>
+): Promise<RehypeExpressiveCodeEngineRenderer<E>> {
 	// Transfer deprecated `theme` option to `themes` without triggering the deprecation warning
-	const deprecatedOptions: Omit<RehypeExpressiveCodeOptions, 'theme'> & { theme?: ThemeObjectOrShikiThemeName | ThemeObjectOrShikiThemeName[] | undefined } = options
+	const deprecatedOptions: Omit<RehypeExpressiveCodeCommonOptions<ExpressiveCodeCoreConfig, E, T>, 'theme'> & {
```

Comment:
Seems to be deprecated over a year ago, is it time to remove completely?  While this PR does not currently make any changes to the current public API, if we do make some minor ones, might be a good time to remove support for `theme`.  Also, since EC is `v0.x`, technically every release is to be considered a potential `breaking change` by package consumers :)

Reply:
Good question. I've just recently come across a user who still used the old `theme` option. Without this encounter, I would have probably agreed. :) I think I'll remove all deprecated features when finally switching to `v1.0`, which I'm planning to do this year. I'm not a fan of the special pre-`v1.0`-semver behavior and will be happy to get out of this zone. :)

Reply:
Sounds good, will leave as-is, closing this as resolved.